price_feed/lib.rs

```diff
@@ -0,0 +1,197 @@
+            ::ink::env::chain_extension::ChainExtensionMethod::build(7777u32)
+                .input::<([u8; 32], [u8; 32])>()
+                .output::<Result<CoinInfo, ChainExtensionError>, false>()
+                .handle_error_code::<ChainExtensionError>()
+                .call(&input)?
+        }
+    }
+
+    #[derive(Debug, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub struct CoinInfo {
+        pub symbol: Vec<u8>,
+        pub name: Vec<u8>,
+        pub blockchain: Vec<u8>,
+        pub supply: u128,
+        pub last_update_timestamp: u64,
+        pub price: u128,
+    }
+
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum ChainExtensionError {
+        /// Some error occurred.
+        Other,
+        /// Failed to lookup some data.
+        CannotLookup,
+        /// A bad origin.
+        BadOrigin,
+        /// A custom error in a module.
+        Module,
+        /// At least one consumer is remaining so the account cannot be destroyed.
+        ConsumerRemaining,
+        /// There are no providers so the account cannot be created.
+        NoProviders,
+        /// There are too many consumers so the account cannot be created.
+        TooManyConsumers,
+        /// An error to do with tokens.
+        Token(PalletAssetTokenErr),
+        /// An arithmetic error.
+        Arithmetic(PalletAssetArithmeticErr),
+        // unknown error
+        Unknown,
+    }
+
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum PalletAssetArithmeticErr {
+        /// Underflow.
+        Underflow,
+        /// Overflow.
+        Overflow,
+        /// Division by zero.
+        DivisionByZero,
+        // unknown error
+        Unknown,
+    }
+
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum PalletAssetTokenErr {
+        /// Funds are unavailable.
+        NoFunds,
+        /// Account that must exist would die.
+        WouldDie,
+        /// Account cannot exist with the funds that would be given.
+        BelowMinimum,
+        /// Account cannot be created.
+        CannotCreate,
+        /// The asset in question is unknown.
+        UnknownAsset,
+        /// Funds exist but are frozen.
+        Frozen,
+        /// Operation is not supported by the asset.
+        Unsupported,
+        // unknown error
+        Unknown,
+    }
+
+    impl ink::env::chain_extension::FromStatusCode for ChainExtensionError {
+        fn from_status_code(status_code: u32) -> Result<(), Self> {
+            match status_code {
+                0 => Ok(()),
+                _ => panic!("encountered unknown status code"),
```

Comment:
Should we panic? Is it normal to panic in a smart contract?

Reply:
In a contract, panic will just interrupt that contract's execution. It won't crash the chain or anything. But yes, we should avoid panic in favor of actual errors. Done 👍